### ISSUE TITLE
Feature Request : Gumbel Softmax

### ISSUE THREAD

**Issue by Sentient07** [Theano/Theano#6039]
The equation (1) in this [paper](https://arxiv.org/pdf/1611.01144.pdf) gives the equation of a sampler that samples from a categorical distributions. The advantage of this discrete sampler over the others is that it is differentiable and could be trained with backprop in an environment of discrete data

**Comment by nouiz** [comment]
Can you put the equation in the issue to help make it easier to follow? I
checked and Theano have all the ops needed to implement it. Mostly, log,
elemwise add, argmax and one_hot. one_hot can be done with sub alloc(0,
shape) and set_subtensor.

**Comment by Sentient07** [comment]
> Can you put the equation in the issue to help make it easier to follow?

I tried to add in the comment field, but the code of github doesn't support latex, so I added a screenshot with the three necessary equations (sampling, argmax approximation, gumbel distribution's density)

>  one_hot can be done with sub alloc(0, shape) and set_subtensor.

I thought this could be done with the [Op](http://deeplearning.net/software/theano/library/tensor/extra_ops.html#theano.tensor.extra_ops.to_one_hot) , but why the longer approach?

**Comment by nouiz** [comment]
The longer to implement approch is using an op. The faster to implement is
using existing ops and build a graph. This can be hidden inside a python
function.

So when do we need a new op? If we want more speed, sometimes, it could be
faster. But it isn't always the case. But this would request you to make a
CPU and GPU version. To implement the grad on CPU an GPU. So this is much
more time consuming. Start with a graph version and only try to make an op
version if this part take a significant execution time.

**Comment by Sentient07** [comment]
> The faster to implement is using existing ops and build a graph.

Ah, I get it! 

> If we want more speed, sometimes, it could be faster. But it isn't always the case.

Basically, if I build a graph to do this computation with the existing Ops and it happens to be costly, then it is worth making a new Op. Am I correct ?

> Start with a graph version and only try to make an op version if this part take a significant execution time.

Makes sense

**Comment by Sentient07** [comment]
This is how the graphs look (first is the that of the variable `z` in equation (1) and second is variable `y` in equation (2)). The crux here is we can use (1) in the forward pass, and differentiate (2) at the backward pass. 

P. S : From what I understand, this approximation of using (1) for forward and (2) in backward holds good at low temperature, but the training they have described in the paper is to start from high temperature and anneal it. 

<img width="731" alt="screenshot at jun 14 16-00-19" src="https://user-images.githubusercontent.com/6195312/27128209-3b98ea82-511b-11e7-9570-8b29ef817ff2.png">
<img width="752" alt="screenshot at jun 14 16-00-34" src="https://user-images.githubusercontent.com/6195312/27128214-3e01d338-511b-11e7-9f23-7270b00064ee.png">

**Comment by nouiz** [comment]
I don't have the time to check the graph in detail, but it seem resonable.
Now benchmark that in a real model.

**Comment by Sentient07** [comment]
Forgot to add the code, here is the [gist](https://gist.github.com/Sentient07/81bdf04de7d8347b429734da0fecbaf6) .

**Comment by f0k** [comment]
>> I thought this could be done with the [Op](http://deeplearning.net/software/theano/library/tensor/extra_ops.html#theano.tensor.extra_ops.to_one_hot), but why the longer approach?

> The longer to implement approch is using an op. The faster to implement is using existing ops and build a graph.

I think this was a misunderstanding -- @Sentient07 was just referring to the "one-hot" part. The `extra_ops.to_one_hot` actually is a Python function doing what @nouiz suggested: a zero tensor with a set_subtensor: https://github.com/Theano/theano/blob/52903f8/theano/tensor/extra_ops.py#L1079-L1104. It's okay to use the existing function.

> I am not sure what is ideal, having this as a Theano Op or Lasagne nonlinearity.

I think the question you actually meant was: Having this in Theano or only in Lasagne.
In both cases, there will be a function like `T.nnet.gumbel_softmax` or `lasagne.nonlinearities.gumbel_softmax` that creates the graph, and we can decide whether to use existing Ops for the graph or create a new Theano Op (with all the implications @nouiz) mentioned above.
Whether to add it to Theano or only to Lasagne is a question of how generally useful this is, and who should maintain it.

> Forgot to add the code, here is the [gist](https://gist.github.com/Sentient07/81bdf04de7d8347b429734da0fecbaf6).

Note that there's also a gumbel-softmax and gumbel-sigmoid implementation by @justheuristic linked at https://github.com/Lasagne/Lasagne/issues/784.

**Comment by Sentient07** [comment]
> Whether to add it to Theano or only to Lasagne is a question of how generally useful this is, and who should maintain it.

Basically it helps in sampling from a categorical distribution or in discrete space and allows you train it using back prop. I would suggest we could have in theano because it's a first of kind(also theano has been maintaining quite a few variations of softmaxes )

> It's okay to use the existing function.

The graphs were the same I realised then 😛 

> Note that there's also a gumbel-softmax and gumbel-sigmoid implementation

I saw this implementation, it has everything that is said in the paper, but I think we need something like an `OpFromGraph`. Because, what we use while computing gradient and what we expect as output are different here. I am working on wrapping it inside an `OpFromGraph` in a way that given the inputs, it will automatically build the softmax (equation `2`) for computing the gradient and one hot encoded argmax as the output.

**Comment by f0k** [comment]
> I saw this implementation, it has everything that is said in the paper, but I think we need something like an `OpFromGraph`.

Only for the ST (Straight-Through) Gumbel-Softmax. So this should be optional.

**Comment by Sentient07** [comment]
> So this should be optional. 

Yeah, I will follow a more general approach.

> We might want to have two parameters (in addition to the class probabilities)

I am not sure if we just need those alone. I am not able to get the OpFromGraph implementation correct. I have this [problem](https://github.com/Theano/Theano/issues/5685) . Also, if we are going to have an OpFromGraph implementation, I think we need to have the epsilon as a `fscalar` and not a constant.

**Comment by f0k** [comment]
> I have this problem.

I think you shouldn't run into this problem because you'll want to use an inline OpFromGraph. I assume that wouldn't be affected. (But I'm not very sure.)
